### ISSUE TITLE
Improve VM's JSON decoder 

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#55522]
We sometimes hear complaints that our JSON decoder is slow compared to, for example, V8's. Usually we just shrug these complaints away saying that:

- Dart's JSON decoder is likely fast enough 
- Dart's JSON decoder is written in Dart and V8's is written in C++ - so some degree of performance difference is expected. 

Developers are then encouraged to use high performance native JSON decoders if JSON decoding is a bottleneck for them. 

This does not square very well with my belief that Dart should be considered a general purpose high-level programming language. So I would like to take a closer look at JSON's decoder performance and see if it can be considerably improved. 

I have taken JSON files from https://github.com/simdjson/simdjson-data/tree/master/jsonexamples and run them through [JS and Dart benchmarks](https://gist.github.com/mraleph/17add65c8e00082a210cb84adca377fd). 

Dart benchmark is measuring performance of JSON decoder fused with UTF8 decoder:

```dart
  final decoder = Utf8Decoder().fuse(JsonDecoder());
  // ...
    for (var i = 0; i < N; i++) {
      decoder.convert(bytes);
    }
```

JS benchmark is measuring performance of decoding JSON from pre-decoded JS string as well as performance of coverting UTF8 bytes to a JS string:

```js
function benchmark(baseDir, fileName) {
  let content = read(/* ... */);
  // ...
    for (var i = 0; i < N; i++) {
      JSON.parse(content);
    }
  // ...
}

function benchmarkWithUtf8Decoding(baseDir, fileName) {
  let content = readbuffer(/* ... */);
    // ...
    for (var i = 0; i < N; i++) {
      JSON.parse(decodeUtf8(content));
    }
    // ...
}
```

`decodeUtf8` is a helper function which I added to `d8` shell which effectively does the following: `String::NewFromUtf8(isolate, buffer, NewStringType::kNormal, buffer_size)`. 

This was done to measure the cost of utf8 -> utf16 convertion which will otherwise be hidden. 

The baseline measurement for V8 on my desktop looks like this:

```
"apache_builds.json",string,0.30905068359375
"apache_builds.json",buffer,0.3241399414062499
"canada.json",string,14.897924999999987
"canada.json",buffer,16.510981250000007
"citm_catalog.json",string,3.759248437500003
"citm_catalog.json",buffer,6.9653749999999945
"github_events.json",string,0.14244892578124996
"github_events.json",buffer,0.18018886718749982
"google_maps_api_compact_response.json",string,0.07089770507812503
"google_maps_api_compact_response.json",buffer,0.07248544921875003
"google_maps_api_response.json",string,0.08099453125
"google_maps_api_response.json",buffer,0.08426477050781252
"gsoc-2018.json",string,4.6573312500000155
"gsoc-2018.json",buffer,6.232159375000014
"instruments.json",string,0.49713339843750076
"instruments.json",buffer,0.635085742187502
"marine_ik.json",string,13.519856249999975
"marine_ik.json",buffer,14.717656249999981
"mesh.json",string,2.9209296875000064
"mesh.json",buffer,3.354901562500004
"mesh.pretty.json",string,4.357474999999999
"mesh.pretty.json",buffer,5.1385874999999945
"numbers.json",string,0.5762896484374977
"numbers.json",buffer,0.6817013671874974
"random.json",string,2.822692968749993
"random.json",buffer,3.569248437499982
"repeat.json",string,0.025887060546875063
"repeat.json",buffer,0.051292358398437446
"semanticscholar-corpus.json",string,50.387549999999464
"semanticscholar-corpus.json",buffer,55.51836249999978
"tree-pretty.json",string,0.09902978515625023
"tree-pretty.json",buffer,0.10347050781250004
"twitter_api_compact_response.json",string,0.03364210205078137
"twitter_api_compact_response.json",buffer,0.04489652099609387
"twitter_api_response.json",string,0.04411997070312523
"twitter_api_response.json",buffer,0.06057795410156288
"twitterescaped.json",string,1.7861546875000158
"twitterescaped.json",buffer,2.1810593749999954
"twitter.json",string,1.7194124999999985
"twitter.json",buffer,3.385740625000017
"twitter_timeline.json",string,0.13089736328124957
"twitter_timeline.json",buffer,0.13675043945312382
"update-center.json",string,2.1542187499999956
"update-center.json",buffer,3.237673437500007
```

Dart in AOT mode yields:

```
"apache_builds.json",0.6998046875
"canada.json",31.8375
"citm_catalog.json",9.1125
"github_events.json",0.39033203125
"google_maps_api_compact_response.json",0.115283203125
"google_maps_api_response.json",0.162744140625
"gsoc-2018.json",20.49375
"instruments.json",1.45390625
"marine_ik.json",23.26875
"mesh.json",4.640625
"mesh.pretty.json",11.571875
"numbers.json",0.570703125
"random.json",4.7703125
"repeat.json",0.0804443359375
"semanticscholar-corpus.json",69.275
"tree-pretty.json",0.22900390625
"twitter_api_compact_response.json",0.0744384765625
"twitter_api_response.json",0.1037109375
"twitterescaped.json",4.553125
"twitter.json",4.6828125
"twitter_timeline.json",0.45234375
"update-center.json",4.2859375
```

With the exceptions of `numbers.json` Dart's decoder is 1.5-3x slower.

Good news is that it is not impossible to significantly improve its performance - I have done some experiments and got to the following preliminary numbers:

```
"apache_builds.json",0.3947265625
"canada.json",30.7375
"citm_catalog.json",5.0671875
"github_events.json",0.22177734375
"google_maps_api_compact_response.json",0.093896484375
"google_maps_api_response.json",0.101416015625
"gsoc-2018.json",8.790625
"instruments.json",0.956640625
"marine_ik.json",18.00625
"mesh.json",3.984375
"mesh.pretty.json",8.625
"numbers.json",0.49609375
"random.json",3.4328125
"repeat.json",0.04671630859375
"semanticscholar-corpus.json",36.4875
"tree-pretty.json",0.14794921875
"twitter_api_compact_response.json",0.056640625
"twitter_api_response.json",0.071142578125
"twitterescaped.json",3.8953125
"twitter.json",3.1515625
"twitter_timeline.json",0.37626953125
"update-center.json",2.89453125
```

Which are significantly better. 

I am going to use this issue as an umbrella issue to land my changes.

**Comment by modulovalue** [comment]
@mraleph Do you see any potential performance improvements from merging the underlying regular UTF8 automaton with an automaton that lexes JSON and using that as the lexer for the JsonDecoder (to have a Utf8JsonDecoder) so that no fusing is needed?

**Comment by lrhn** [comment]
The fusing creates a [UTF-8 specialized JSON decoder](https://github.com/dart-lang/sdk/blob/5b2ea0c7a227d91c691d2ff8cbbeb5f7f86afdb9/sdk/lib/_internal/vm/lib/convert_patch.dart#L41), `_JsonUtf8Decoder`.

**Comment by modulovalue** [comment]
Thank you for the hint @lrhn. It looks to me like the specialized UTF8-JSON decoder is using a table-driven UTF8 decoder and a non-table-driven JSON parser.

What I meant to say [in this comment](https://github.com/dart-lang/sdk/issues/55522#issuecomment-2080446091) was, since the language of UTF8 appears to be regular, and the lexical structure of JSON seems to be regular too, it feels like there might be performance gains to be had by merging both automata into one and having the tokenization phase be based on a single automaton that does UTF8 decoding and JSON lexing at the same time.

**Comment by lrhn** [comment]
I think the UTF-8 is a red herring here. 
A table based tokenizer would work equally well for both UTF-8 and string based JSON parsing, since everything outside of string values is ASCII anyway, and inside strings need to be interpreted it contains escapes -- or non-ASCII characters in a UTF-8 source.

The reason the current implementation doesn't use a table based tokenizer is that it doesn't need to tokenize at all. The moment it sees the first character of a value, it can start interpreting it. It tries to collect the value of a number and the contents of a string _while_ scanning, in one pass, so it doesn't have to first tokenize, and then interpret the token afterwards.

Maybe that process can be made more efficient, but doing an extra tokenization step first doesn't sound like an improvement. 
(Unless it can be done very efficiently using SIMD operations or using parallelism, or something.)

**Comment by rakudrama** [comment]
The current decoder generates maps that take a lot of memory.
Decoding
```
[{"x":1,"y":2},{"x":3,"y":4}]
```
creates fresh strings for the repeated `"x"` (and `"y"`), and these strings are inserted into similar Maps that have their own index.

There is a cost to using a fresh string as a key. Other than memory footprint, the `hashCode` needs to be computed for each instance. Perhaps this cost could be better used to identify a common index to be shared by similar JSON objects (marked as copy-on-write if used more than once).

**Comment by lrhn** [comment]
Maps and strings can _probably_ be optimized.
Large JSON inputs trendy to have repetitions of the same map structure. 
Canonicalized map keys, maybe even some short values, can avoid doing as many allocations, and if it's possible to reuse the map structure, then there might be a saving there too.
The cost of that is an overhead while using, for recognizing those existing values and structures. I have tried it before, and the overhead was visible. 

It's much more efficient to build a custom parser for the known structures using a more pull-based JSON interpreter, than to try to recognize possibly repeated structures at parse time in a generic parser.

(That is: if you care about performance, why are you building maps and lists at all?)

**Comment by mraleph** [comment]
> The current decoder generates maps that take a lot of memory.

I have a prototype patch which canonicalizes keys when decoding JSON (this helps both with memory consumption and improves access speed as well, because accessing maps with literal keys can short-circuit string comparison later). You still need to compute `hashCode` though. 

I entertained the idea of using "predictive" decoding, where decoding an array of JSON objects makes an assumption about repeated shape and uses that to short-circuit both key decoding and `Map` hashing and maybe even use specialized COW `Map` implementation which allows sharing of the "index" part. This is what V8 JSON decoder effectively does. But I am not sure I want to go there - I feel the need to keep JSON decoder complexity down. My thinking here is aligned with @lrhn's: if code is doing `bytes -> JSON objects (`Map`'s) -> proper Dart objects` then it is better to actually do `bytes -> proper Dart objects`.

**Comment by gnprice** [comment]
A pull-based JSON interpreter for going straight from bytes to proper Dart objects sounds great. Previous issue threads for that:
* https://github.com/dart-lang/sdk/issues/35693
* https://github.com/google/json_serializable.dart/issues/301

**Comment by kevmoo** [comment]
See also https://github.com/kevmoo/json_alt – although this code is VERY old

**Comment by schultek** [comment]
See also https://github.com/simc/crimson

**Comment by feinstein** [comment]
Maybe we should look also at how other languages have dealt with this? I feel that if we only look at how js and dart solutions have being built so far, we might miss some clever solutions that might exist for other languages.

**Comment by AdamTechAcc** [comment]
> See also https://github.com/simc/crimson

Someone uses it?

**Comment by modulovalue** [comment]
The story here is sadly very unfortunate right now (/cc @mit-mit).

I've benchmarked a generated direct coded JSON lexer in Dart against 9 other languages (C, Swift, Java, Go, ...). All generated from the same DFA and using practically the same code generation pattern. **Dart is unfortunately the slowest out of all of them by a significant amount**.

_Note that table driven lexers are intrinsically slower so they are not an option._

My tests show that a generated lexer + parser (included Map building overhead) is essentially as fast as the native JSON decoder, so I think there's not much that can be done on user side here.

I've provided a repro here: https://github.com/modulovalue/dart-lexing-regression (also see: https://github.com/dart-lang/sdk/issues/63230). It contains generated lexers for 3 different languages (JSON, Rust, LLVM-IR) in C and Dart, each an order of magnitude greater in size.

@mraleph any ideas what could be slowing us down? Any help would be appreciated!

**Comment by lrhn** [comment]
JSON is a very simple, but very verbose, format.
The big cost of JSON parsing isn't the _parsing_, it's simple enough that you can do that with a one-pass zero-lookahead pass.
The cost is the allocation of all those darn objects. It's like it's an object model designed for a language that never cared about performance!
And for every list or map you build while parsing, you don't know the size ahead of time without scanning ahead, so allocation tend to be growing lists and maps.
You can avoid some of that, but it costs as an overhead even when it doesn't help.

(Maybe we should heuristically use a simple parser on short inputs, and an allocation-optimizing parser for large inputs.)

**Comment by gnprice** [comment]
I think in most applications the lists and maps produced by the JSON decoder are ephemeral, getting immediately consumed to produce proper Dart objects (e.g. by a `fromJson` method generated by `package:json_serializable`).

So for cutting out that overhead, I like the idea discussed above and at #35693 of a pull-based JSON reader that can be used to build the intended objects directly.

**Comment by modulovalue** [comment]
@lrhn said:

> The cost is the allocation of all those darn objects

Citation needed! 😀 

Here's a quick benchmark I threw together:

```
aggregated over 30 files (23.06 MB), correctness vs jsonDecode: all pass

step                                                                time      % pipeline
----------------------------------------------------------------------------------------
1. bytes -> tokens                                             63.066 ms          42.9 %
2. tokens -> events                                            16.624 ms          11.3 %
3. events -> Map/List object tree                              67.345 ms          45.8 %
----------------------------------------------------------------------------------------
our pipeline (sum of steps)                                   147.035 ms

dart:convert jsonDecode(String)                                73.214 ms
dart:convert utf8+json fused (bytes -> object)                 55.698 ms
pipeline / jsonDecode(String)                                      2.01x
pipeline / utf8+json fused (apples-to-apples)                      2.64x


step 3 breakdown (substep run in isolation across whole corpus):
substep                                                             time         vs walk
----------------------------------------------------------------------------------------
walk only (iterate events, no allocation)                       0.604 ms       +0.000 ms
structure only (Map/List alloc)                                12.836 ms      +12.232 ms
strings only (decode every key/string, discard)                25.369 ms      +24.765 ms
numbers only (decode every number, discard)                    18.092 ms      +17.488 ms
full (current step 3)                                          62.080 ms      +61.476 ms
```

Even if we cut the performance of the lexical analysis step in half, its still on the critical path and [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl's_law) says it's still worth optimizing. (As I said, out of the 10 languages I tested, **Dart is by far the slowest when it comes to lexing!**)

@lrhn said:

> The big cost of JSON parsing isn't the parsing

Unfortunately, "parsing" is overloaded to either mean lexing alone or parsing and lexing at the same time (i.e. a scannerless parser), so let's call it lexical analysis and structural analysis to be precise. I think you are referring to structural analysis as being fast and I completely agree! (see above, 7%, not worth optimizing). However, I was not referring to the structural analysis as being slow, but the lexical analysis being slow. And since all steps need to happen sequentially (even if they are fused), optimizing the lexical analysis use case is a high value endeavour and possibly the least effort one for making jsonDecode faster.
 
---

@gnprice said:
> You can avoid some of that, but it costs as an overhead even when it doesn't help.

There are two separate discussions to be had when it comes to json performance:
 - parsing json in Dart (via JsonDecoder) is slow: how do we make JsonDecoder faster?
 - there are alternative APIs that would support a more efficient parser: can we have a new API to parse json faster that way?

Deferring work into the future by, for example, supporting laziness in some form, alternative object models, cursor APIs (a la tree sitter) does not make _our_ `JsonDecoder` faster, so I'd have to point out that this is somewhat off topic since this issue focuses on making JsonDecoder faster. The API is eager changing that would be a major breaking change.

**Comment by kevmoo** [comment]
FYI: on the ENCODE side, I have some great results: https://github.com/kevmoo/json_serializable.dart/blob/rw/json_rw/docs/benchmark_results.md

**Comment by kevmoo** [comment]
This solution uses a reader/writer model.

**Comment by modulovalue** [comment]
Lasse said:

> The fusing creates a [UTF-8 specialized JSON decoder](https://github.com/dart-lang/sdk/blob/5b2ea0c7a227d91c691d2ff8cbbeb5f7f86afdb9/sdk/lib/_internal/vm/lib/convert_patch.dart#L41), _JsonUtf8Decoder.

The utf-8 specialized JSON decoder is significantly faster than jsonEncode. It looks like we might be having a UX problem here that is very easy to solve.

[Stack Overflow](https://stackoverflow.com/questions/15866290/the-best-way-to-parse-a-json-in-dart), Claude, Gemini, ChatGPT all recommend jsonEncode, which takes a String. The fast path is essentially hidden due to its verbosity: `utf8.decoder.fuse(jsonDecoder).convert(source)`.

We should have:

```dart
  /// Parses [source] (UTF-8 encoded JSON bytes) and returns the resulting
  /// JSON object.
  ///
  /// Equivalent to `jsonDecode(utf8.decode(source), reviver: reviver)`,
  /// but the bytes are scanned directly by the SDK's UTF-8 JSON parser
  /// instead of being decoded to a [String] first. Typical speedup is
  /// 1.5x to 3x on real-world JSON workloads, since most of JSON's
  /// syntax is single-byte ASCII and only string-literal contents need
  /// any UTF-8 decoding.
  ///
  /// The optional [reviver] function is called once for each object or
  /// list property that has been parsed during decoding. The `key`
  /// argument is either the integer list index for a list property, the
  /// string map key for object properties, or `null` for the final result.
  ///
  /// The default [reviver] (when not provided) is the identity function.
  ///
  /// Throws [FormatException] on invalid JSON or invalid UTF-8 inside a
  /// string literal (the latter would be invalid JSON anyway).
  ///
  /// Example:
  /// ```dart
  /// final bytes = await File('data.json').readAsBytes();
  ///
  /// final data = jsonDecodeBytes(bytes);
  /// print(data['text']);   // foo
  /// print(data['value']);  // 1
  /// print(data['status']); // false
  /// print(data['extra']);  // null
  /// ```
  dynamic jsonDecodeBytes(
    List<int> source, {
    Object? Function(Object? key, Object? value)? reviver,
  }) {
    final jsonDecoder = reviver == null ? json.decoder : JsonDecoder(reviver);
    return utf8.decoder.fuse(jsonDecoder).convert(source);
  }
```

and maybe a File.readAsJson + a lint to warn about readAsString + jsonDecode. The easy case, two consecutive statements, are easy to detect and should cover most cases.

@lrhn what do you think?

**Comment by natebosch** [comment]
I'm skeptical of `File.readAsJson` but I think I'm in favor of having some API targeting `List<int>` if we can provide an implementation that most would prefer over `jsonDecode`.

**Comment by lrhn** [comment]
Definitely worth it to not allocate the string. Non-ASCII UTF-8 only occurs in string literals, which also need to handle escapes anyway.
We should just have a top-level `utf8Json` codec, for discoverability. If we have both the UTF-8 decoder and encoder already, it shouldn't be much work.

As usual, if you care about speed to begin with, _why are you allocating maps and lists_?

Imagine if:
* The maps and list generated by JSON decoding were not mutable. (You could always clone them if you wanted to work on them. Or maybe we could make them copy-on-write.)
    * Heck, if all the values are strings, you could put them in a `List<String>` or `Map<String, String>`, instead of using `dynamic`.
* Map keys are canonicalized. String values too if they're _small_ strings. Maps with the same key (sequence) use the same hash table. (Basically V8's object map idea, including map transitions.)
* Map and list values are parsed lazily. The limits of the values are found while scanning, but the actual values are lazily constructed. (Optional, if you know you'll traverse everything anyway, might as well create it up front.)
* Strings are created lazily. The map key representation doesn't need to be strings, they can be something that can be efficiently compared to a string. That can be a source range in the `Uint8List` (pre-scanned to know that it's ASCII or plain UTF-8, if it contains escapes, it probably needs to be re-represented, a single index into a canonicalized table, heck even the bytes of short ASCII strings stored directly in a tagged integer. Only create the string if needed.
* Or there was a way to iterate/event-based-scan the JSON directly as UTF-8, so you can choose to parse a string only if you need it, and can compare keys directly against the source ranges. (Like `package:json_tool` tries to do.)

For AI contexts, which keep getting shuffled back and forth, the client shouldn't decode the context every time and then recode it to send it back. Keep everything important as UTF-8 in a persistent buffer, or the buffers it came in originally, and copy it directly into the network frames.

Or if the the JSON scanner can clobber the input bytes, it can re-build a binary JSON representation structure directly in those bytes, instead of working on the raw bytes.

**Comment by mraleph** [comment]
@modulovalue I have spent some time in 2024 improving performance of fused JSON decoder and gave a talk about it on Flutter'n'Friends. Check out the [slides](https://mrale.ph/talks/fnf2024/) and corresponding [video](https://www.youtube.com/watch?v=CbvRT9HObOw). I was fairly satisfied with end results - though there were still some outliers left where our JSON decoder was 2x slower than V8's, though you can make more improvements within the current paradigm, thinks that I can remember from the top of my head: 

- parsing strings with escapes is especially bad right now;
- utf8 decoding can be improved
- parsing lists of maps with repeating shapes can be improved
- doing more parsing recursively rather than in a streaming fashion can help

But if performance of JSON parsing is truly important to you then switching away from builtin JSON parser is even better. https://github.com/mraleph/dart_simdjson was a demo showing how you can build a fast parser on top of `simdjson` library.

One improvement we could make is to make a version of df80cf91404e8e3b0f0a4eb271467448d126199e for `jsonDecode` - specialize internal implementation of JSON parsing for one-byte and two-byte strings, to make character access fast. I think  that alone will speed up `jsonDecode` by 30-40%.

**Comment by modulovalue** [comment]
I did some more investigating and managed to port simdjson to Dart (not ffi, direct port using first class SIMD operations).

```
                                             (in MB/s)
  corpus                          size MB    stage1    stage2    sdk simdjson    jsonDecode    jsonDecodeBytes
                                              (SIMD)    (SDK)    (stg1+stg2)    (String in)    (fused decoder)
  apache_builds.json                 0.12      1485       463            412           473                508
  canada.json                        2.15      3739       194            171           195                186
  citm_catalog.json                  1.65      3750      1107            688           428                586
  github_events.json                 0.06      1487       804            632           482                589
  google_maps_api_response.json      0.02      2940       790            581           416                460
  gsoc-2018.json                     3.17      5123       580            465           634                706
  instruments.json                   0.21      3102       756            382           417                482
  marine_ik.json                     2.85      2585       502            450           307                340
  mesh.json                          0.69      3503       504            390           374                423
  mesh.pretty.json                   1.50      5166       426            352           403                408
  numbers.json                       0.14      4211       953            435           646                740
  random.json                        0.49      1859       393            232           198                309
  repeat.json                        0.01      3402       574            482           379                404
  twitter.json                       0.60      3469       594            373           307                459
  update-center.json                 0.51      2400       569            329           219                452
```

## simdjson stage 1

Stage 1 consists of several phases where each takes 10-40GB/s. The output is a typed buffer of structural symbols. The good news is that Dart can keep up with C (and even beat it in some of those phases!). Some optimization passes are missing in the VM for Dart to be able to reach C performance numbers, but Dart (both JIT and AOT) is absolutely capable of getting solid numbers in Stage 1. I don't see anything preventing us from being able to go head to head with C or other systems programming languages here, which is amazing!

## simdjson stage 2

The structural symbols from stage 1 are parsed into Dart Objects and [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law) bites us (well, me, because this shows Lasse was absolutely correct!), that is, building the data structures is on the critical path and that slows us down, as can be seen in the numbers above. simdjson's own second stage produces a [tape](https://simdjson.org/api/0.4.0/md_doc_tape.html) which is a list of events that can be lazily interpreted into the target data structures. This is not compatible with our interface so we can discard the tape based approach for this issue.


## Regarding jsonDecodeBytes

I benchmarked jsonDecode vs our fused json decoder and while the fused json decoder is noticeably faster for some datasets, there are some datasets where it performs slightly worse. I no longer think we should introduce a jsonDecodeBytes for the current fused decoder so that we keep the option open to change the runtime type to a more specialized version in the future if needed (e.g., List<num> -> Float64List, List<String> -> StringList, List<dynamic> -> DynamicList, Map<String, dynamic> -> StringMapOfDynamics, ...) and changing the runtime type would be breaking then if we were to introduce a jsonDecodeBytes now and change it in the future.

---

TLDR: to make json decoding reach state of the art numbers in Dart we need to tackle this from 2 directions:

 - we need better SIMD support if we want to reach multi GB/s numbers. The Wasm SIMD API should be enough for that. The existing VM infrastructure provides everything that's needed for that to be supported well. (@mraleph get ready for some more CLs!)
 - we need new APIs for json. I don't have strong opinions on that and would love to hear @lrhn's thoughts on [#35693](https://github.com/dart-lang/sdk/issues/35693) since Dart has changed a lot since 2020, maybe you have some new ideas for a modern json API?

**Comment by kevmoo** [comment]
We'd talked about having a "Codable" model. (En-Codable and De-Codable).

An API like this: https://github.com/kevmoo/json_serializable.dart/blob/rw/json_rw/lib/src/json_reader.dart

I'd love to figure out how we add the smallest possible "atoms" to the SDK and do most of the API definition OUTSIDE the SDK.

Looks like there are more than a few people who have looked at this!

https://github.com/schultek/codable/blob/main/lib/src/core/encoder.dart

I think @jakemac53 had an idea about this, too. Not sure where his code is.
